pyquil/quil.py

```diff
@@ -981,3 +983,36 @@ def percolate_declares(program: Program) -> Program:
     p._defined_gates = program._defined_gates
 
     return p
+
+
+def validate_protoquil(program: Program) -> None:
+    """
+    Ensure that a program is valid ProtoQuil, otherwise raise a ValueError.
+    Protoquil allows a global RESET before any gates, and MEASUREs on each qubit after any gates
+    on that qubit. Pragmas are always allowed, and a final Halt instruction is allowed.
+
+    :param program: The Quil program to validate.
+    """
+    gates_seen = False
+    halted = False
+    measured_qubits = set()
+    for instr in program.instructions:
+        if isinstance(instr, Pragma) or isinstance(instr, Declare):
+            continue
+        elif isinstance(instr, Halt):
+            halted = True
+        elif halted:
+            raise ValueError(f"Cannot have instruction {instr} after HALT")
+        elif isinstance(instr, Gate):
+            gates_seen = True
+            if any(q.index in measured_qubits for q in instr.qubits):
+                raise ValueError("Cannot apply gates to qubits that were already measured.")
+        elif isinstance(instr, Reset):
+            if gates_seen:
+                raise ValueError("ProtoQuil disallows RESET after a gate application.")
+        elif isinstance(instr, ResetQubit):
+            raise ValueError("ProtoQuil only allows for global RESET.")
+        elif isinstance(instr, Measurement):
+            measured_qubits.add(instr.qubit.index)
+        else:
+            raise ValueError(f"Unhandled instruction type in ProtoQuil validation: {instr}")
```

Comment:
this will print the string representation of the instruction, rather than its type